filament/backend/src/vulkan/caching/VulkanDescriptorSet.cpp

```diff
@@ -0,0 +1,646 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "VulkanDescriptorSet.h"
+
+#include <vulkan/VulkanConstants.h>
+#include <vulkan/VulkanResources.h>
+
+#include <type_traits>
+
+namespace filament::backend {
+
+namespace {
+
+template<typename MaskType>
+using LayoutMap = tsl::robin_map<MaskType, VkDescriptorSetLayout>;
+using LayoutArray = VulkanDescriptorSetManager::LayoutArray;
+using UniformBufferBitmask = VulkanDescriptorSetManager::UniformBufferBitmask;
+using SamplerBitmask = VulkanDescriptorSetManager::SamplerBitmask;
+
+constexpr uint8_t EXPECTED_IN_FLIGHT_FRAMES = 3; // Asssume triple buffering
+
+#define PORT_CONSTANT(NameSpace, K) constexpr decltype(NameSpace::K) K = NameSpace::K;
+
+PORT_CONSTANT(VulkanDescriptorSetManager, MAX_SUPPORTED_SHADER_STAGE)
+PORT_CONSTANT(VulkanDescriptorSetManager, VERTEX_STAGE)
+PORT_CONSTANT(VulkanDescriptorSetManager, FRAGMENT_STAGE)
+
+PORT_CONSTANT(Program, UNIFORM_BINDING_COUNT)
+PORT_CONSTANT(Program, SAMPLER_BINDING_COUNT)
+
+#undef PORT_CONSTANT
+
+uint8_t BIT_COUNT[256] = {0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4, 1, 2, 2, 3, 2, 3, 3, 4, 2,
```

Comment:
I wonder if we could generate this with a `constexpr` function

Reply:
yes!  thanks for the suggestion.  it reads cleaner.